### DISCUSSION TITLE
Annotate returning it self and Interfaces

### DISCUSSION THREAD

**Discussion by coolyrat** [uber-go/fx#1079] [Q&A]
I have the code like this

```go
var Module = fx.Options(
	fx.Provide(
		New,
		fx.Annotate(New, fx.As(new(Foo)), fx.As(new(Bar))),
	),
)

type Concrete struct{}

func New() *Concrete {
	return &Concrete{}
}

func (c *Concrete) F() {}

func (c *Concrete) B() {}

type Foo interface {
	F()
}

type Bar interface {
	B()
}
```

I want to provide `*Concrete`, `Foo`, `Bar`, the code above could achieve that, but by calling `New` twice, is there a way to provide the Interfaces and it self?

Or

to prevent `New` calling twice, I could use fx.Out, and do something like
```go
type Result struct {
	fx.Out
	C          *Concrete
	Foo      Foo
	Bar      Bar
}

return Result{
		C:     c,
		Foo: c,
		Bar:  c,
	}
```

**Comment by sywhang** [comment]
There is currently no way to do this in Fx. You have to provide a type itself as well as the annotated ones to be able to provide it as all 3 types. 

This could be a feature we could add to fx.Annotate.
